Gradle configuration quarkusDev does not work since 3.25.1

### Describe the bug

Gradle configuration quarkusDev does not work since 3.25.1, after merging this PR #49224

```
dependencies {
    implementation(enforcedPlatform(libs.quarkus.bom))
    implementation(libs.quarkus.rest.jackson)
    quarkusDev(libs.quarkiverse.barcode.zxing)
    compileOnly(libs.quarkiverse.barcode.zxing)
}
```

I got this error:

```
> Task :quarkusDev FAILED

[Incubating] Problems report is available at: file:///C:/Users/fouad/IdeaProjects/quarkus-gradle-dev/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':quarkusDev'.
> Could not resolve all dependencies for configuration ':quarkusDevCompileOnlyConfiguration'.
   > Could not resolve io.quarkus:quarkus-arc:3.15.6.
     Required by:
         root project :
      > Cannot choose between the available variants of io.quarkus:quarkus-arc:3.26.0.CR1:
          - quarkus.dev.conditional-dependency.quarkus-gradle-dev
          - quarkusDevIterativeQuarkus-gradle-devQuarkus-arc5
          - runtime
        All of them match the consumer attributes:
          - Variant 'quarkus.dev.conditional-dependency.quarkus-gradle-dev' capability 'io.quarkus:quarkus-arc:3.26.0.CR1':
              - Unmatched attributes:
                  - Provides org.gradle.category 'library' but the consumer didn't ask for it
                  - Provides org.gradle.libraryelements 'jar' but the consumer didn't ask for it
                  - Provides org.gradle.status 'release' but the consumer didn't ask for it
                  - Provides org.gradle.usage 'java-runtime' but the consumer didn't ask for it
                  - Provides quarkus.dev.conditional-dependency.quarkus-gradle-dev 'on' but the consumer didn't ask for it
          - Variant 'quarkusDevIterativeQuarkus-gradle-devQuarkus-arc5' capability 'io.quarkus:quarkus-arc:3.26.0.CR1':
              - Unmatched attributes:
                  - Provides org.gradle.category 'library' but the consumer didn't ask for it
                  - Provides org.gradle.libraryelements 'jar' but the consumer didn't ask for it
                  - Provides org.gradle.status 'release' but the consumer didn't ask for it
                  - Provides org.gradle.usage 'java-runtime' but the consumer didn't ask for it
                  - Provides quarkusDevIterativeQuarkus-gradle-devQuarkus-arc5 'on' but the consumer didn't ask for it
          - Variant 'runtime' capability 'io.quarkus:quarkus-arc:3.26.0.CR1':
              - Unmatched attributes:
                  - Provides org.gradle.category 'library' but the consumer didn't ask for it
                  - Provides org.gradle.libraryelements 'jar' but the consumer didn't ask for it
                  - Provides org.gradle.status 'release' but the consumer didn't ask for it
                  - Provides org.gradle.usage 'java-runtime' but the consumer didn't ask for it
   > Could not resolve io.quarkus:quarkus-arc:3.15.6.
     Required by:
         root project : > io.quarkiverse.barcode:quarkus-zxing:1.0.1
      > Cannot choose between the available variants of io.quarkus:quarkus-arc:3.26.0.CR1:
          - quarkus.dev.conditional-dependency.quarkus-gradle-dev
          - quarkusDevIterativeQuarkus-gradle-devQuarkus-arc5
          - runtime
        All of them match the consumer attributes:
          - Variant 'quarkus.dev.conditional-dependency.quarkus-gradle-dev' capability 'io.quarkus:quarkus-arc:3.26.0.CR1':
              - Unmatched attributes:
                  - Provides org.gradle.category 'library' but the consumer didn't ask for it
                  - Provides org.gradle.libraryelements 'jar' but the consumer didn't ask for it
                  - Provides org.gradle.status 'release' but the consumer didn't ask for it
                  - Provides org.gradle.usage 'java-runtime' but the consumer didn't ask for it
                  - Provides quarkus.dev.conditional-dependency.quarkus-gradle-dev 'on' but the consumer didn't ask for it
          - Variant 'quarkusDevIterativeQuarkus-gradle-devQuarkus-arc5' capability 'io.quarkus:quarkus-arc:3.26.0.CR1':
              - Unmatched attributes:
                  - Provides org.gradle.category 'library' but the consumer didn't ask for it
                  - Provides org.gradle.libraryelements 'jar' but the consumer didn't ask for it
                  - Provides org.gradle.status 'release' but the consumer didn't ask for it
                  - Provides org.gradle.usage 'java-runtime' but the consumer didn't ask for it
                  - Provides quarkusDevIterativeQuarkus-gradle-devQuarkus-arc5 'on' but the consumer didn't ask for it
          - Variant 'runtime' capability 'io.quarkus:quarkus-arc:3.26.0.CR1':
              - Unmatched attributes:
                  - Provides org.gradle.category 'library' but the consumer didn't ask for it
                  - Provides org.gradle.libraryelements 'jar' but the consumer didn't ask for it
                  - Provides org.gradle.status 'release' but the consumer didn't ask for it
                  - Provides org.gradle.usage 'java-runtime' but the consumer didn't ask for it
> There is 1 more failure with an identical cause.

* Try:
> Ambiguity errors are explained in more detail at https://docs.gradle.org/8.14/userguide/variant_model.html#sub:variant-ambiguity.
> Review the variant matching algorithm at https://docs.gradle.org/8.14/userguide/variant_attributes.html#sec:abm_algorithm.
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 5s
```

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

https://github.com/Eng-Fouad/quarkus-gradle-dev

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

Workaround:

```
disableQuarkusComponentVariants=true
```

--- a/devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/dependency/ApplicationDeploymentClasspathBuilder.java
+++ b/devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/dependency/ApplicationDeploymentClasspathBuilder.java
@@ -354,6 +354,9 @@ public class ApplicationDeploymentClasspathBuilder {
                         project.getConfigurations().getByName(JavaPlugin.COMPILE_ONLY_CONFIGURATION_NAME));
                 config.shouldResolveConsistentlyWith(getDeploymentConfiguration());
                 config.setCanBeConsumed(false);
+                if (!isDisableComponentVariants(project)) {
+                    QuarkusComponentVariants.setCommonAttributes(config.getAttributes(), project.getObjects());
+                }
             });
         }
     }

--- a/integration-tests/gradle/src/main/resources/compile-only-extension-dependency/build.gradle
+++ b/integration-tests/gradle/src/main/resources/compile-only-extension-dependency/build.gradle
@@ -1,0 +1,40 @@
+plugins {
+    id 'java'
+    id 'io.quarkus'
+    id 'application'
+}
+
+repositories {
+    mavenLocal {
+        content {
+            includeGroupByRegex 'io.quarkus.*'
+            includeGroup 'org.hibernate.orm'
+        }
+    }
+    mavenCentral()
+}
+
+dependencies {
+    implementation enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}")
+    implementation 'io.quarkus:quarkus-rest-jackson'
+    compileOnly 'io.quarkus:quarkus-rest' // this is just to make sure quarkusDevCompileOnlyConfiguration can be resolved with proper attributes set
+}
+
+compileJava {
+    options.compilerArgs << '-parameters'
+}
+
+application {
+    mainClass = 'org.acme.EntryPoint'
+}
+
+run {
+    // propagate the custom local maven repo, in case it's configured
+    if (System.properties.containsKey('maven.repo.local')) {
+        systemProperty 'maven.repo.local', System.properties.get('maven.repo.local')
+    }
+}
+
+test {
+    systemProperty "java.util.logging.manager", "org.jboss.logmanager.LogManager"
+}

--- a/integration-tests/gradle/src/main/resources/compile-only-extension-dependency/gradle.properties
+++ b/integration-tests/gradle/src/main/resources/compile-only-extension-dependency/gradle.properties
@@ -1,0 +1,2 @@
+quarkusPlatformArtifactId=quarkus-bom
+quarkusPlatformGroupId=io.quarkus

--- a/integration-tests/gradle/src/main/resources/compile-only-extension-dependency/settings.gradle
+++ b/integration-tests/gradle/src/main/resources/compile-only-extension-dependency/settings.gradle
@@ -1,0 +1,16 @@
+pluginManagement {
+    repositories {
+        mavenLocal {
+            content {
+                includeGroupByRegex 'io.quarkus.*'
+                includeGroup 'org.hibernate.orm'
+            }
+        }
+        mavenCentral()
+        gradlePluginPortal()
+    }
+    plugins {
+      id 'io.quarkus' version "${quarkusPluginVersion}"
+    }
+}
+rootProject.name='code-with-quarkus'

--- a/integration-tests/gradle/src/main/resources/compile-only-extension-dependency/src/main/java/org/acme/EntryPoint.java
+++ b/integration-tests/gradle/src/main/resources/compile-only-extension-dependency/src/main/java/org/acme/EntryPoint.java
@@ -1,0 +1,11 @@
+package org.acme;
+
+import io.quarkus.runtime.Quarkus;
+import io.quarkus.runtime.annotations.QuarkusMain;
+
+@QuarkusMain
+public class EntryPoint {
+    public static void main(String[] args) {
+        Quarkus.run();
+    }
+}

--- a/integration-tests/gradle/src/main/resources/compile-only-extension-dependency/src/main/java/org/acme/ExampleResource.java
+++ b/integration-tests/gradle/src/main/resources/compile-only-extension-dependency/src/main/java/org/acme/ExampleResource.java
@@ -1,0 +1,16 @@
+package org.acme;
+
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.core.MediaType;
+
+@Path("/hello")
+public class ExampleResource {
+
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    public String hello() {
+        return "hello";
+    }
+}

--- a/integration-tests/gradle/src/test/java/io/quarkus/gradle/devmode/CompileOnlyExtensionDependencyDevModeTest.java
+++ b/integration-tests/gradle/src/test/java/io/quarkus/gradle/devmode/CompileOnlyExtensionDependencyDevModeTest.java
@@ -1,0 +1,34 @@
+package io.quarkus.gradle.devmode;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.UUID;
+
+import org.junit.jupiter.api.condition.DisabledOnOs;
+import org.junit.jupiter.api.condition.OS;
+
+import com.google.common.collect.ImmutableMap;
+
+@DisabledOnOs(OS.WINDOWS)
+public class CompileOnlyExtensionDependencyDevModeTest extends QuarkusDevGradleTestBase {
+
+    @Override
+    protected String projectDirectoryName() {
+        return "compile-only-extension-dependency";
+    }
+
+    @Override
+    protected String[] buildArguments() {
+        return new String[] { "clean", "run" };
+    }
+
+    protected void testDevMode() throws Exception {
+        assertThat(getHttpResponse("/hello")).contains("hello");
+
+        final String uuid = UUID.randomUUID().toString();
+        replace("src/main/java/org/acme/ExampleResource.java",
+                ImmutableMap.of("return \"hello\";", "return \"" + uuid + "\";"));
+
+        assertUpdatedResponseContains("/hello", uuid);
+    }
+}

